src/java/org/apache/cassandra/service/MigrationCoordinator.java

```diff
@@ -0,0 +1,512 @@
+                    try
+                    {
+                        mergeSchemaFrom(endpoint, mutations);
+                    }
+                    catch (Exception e)
+                    {
+                        logger.error(String.format("Unable to merge schema from %s", endpoint), e);
+                        return fail();
+                    }
+                }
+                return pullComplete(endpoint, info, true);
+            }
+        }
+
+        public boolean isLatencyForSnitch()
+        {
+            return false;
+        }
+    }
+
+    private void pullSchema(Callback callback)
+    {
+        if (!isAlive(callback.endpoint))
+        {
+            noSpamLogger.warn("Can't send schema pull request: node {} is down.", callback.endpoint);
+            callback.fail();
+            return;
+        }
+
+        // There is a chance that quite some time could have passed between now and the MM#maybeScheduleSchemaPull(),
+        // potentially enough for the endpoint node to restart - which is an issue if it does restart upgraded, with
+        // a higher major.
+        if (!shouldPullFromEndpoint(callback.endpoint))
+        {
+            logger.info("Skipped sending a migration request: node {} has a higher major version now.", callback.endpoint);
+            callback.fail();
+            return;
+        }
+
+        logger.debug("Requesting schema from {}", callback.endpoint);
+        sendMigrationMessage(callback);
+    }
+
+    protected void sendMigrationMessage(Callback callback)
+    {
+        MessageOut<?> message = new MessageOut<>(MessagingService.Verb.MIGRATION_REQUEST, null, MigrationManager.MigrationsSerializer.instance);
+        logger.info("Sending schema pull request to {} at {} with timeout {}", callback.endpoint, System.currentTimeMillis(), message.getTimeout());
+        MessagingService.instance().sendRR(message, callback.endpoint, callback, message.getTimeout(), true);
+    }
+
+    private synchronized Future<Void> pullComplete(InetAddress endpoint, VersionInfo info, boolean wasSuccessful)
+    {
+        if (wasSuccessful)
+            info.markReceived();
+
+        info.outstandingRequests.remove(endpoint);
+        info.requestQueue.add(endpoint);
+        return maybePullSchema(info);
+    }
+
+    /**
+     * Wait until we've received schema responses for all versions we're aware of
+     * @param waitMillis
+     * @return true if response for all schemas were received, false if we timed out waiting
+     */
+    public boolean awaitSchemaRequests(long waitMillis)
+    {
+        if (!FBUtilities.getBroadcastAddress().equals(InetAddress.getLoopbackAddress()))
+            CassandraDaemon.waitForGossipToSettle();
+
+        WaitQueue.Signal signal = null;
+        try
+        {
+            synchronized (this)
+            {
+                List<WaitQueue.Signal> signalList = new ArrayList<>(versionInfo.size());
+                for (VersionInfo version : versionInfo.values())
```

Comment:
Could there be a scenario where other nodes can't reach us within the schema delay wait time and we never populate the `versionInfo` map?